[mlir][gpu]add AffineScope Trait to gpu.launch.

linuxlonelyeagle opened the pull request:
as the title.
* `affine-loop-unroll` works on `func.func`, so you can write the kernel in `gpu.launch`, unroll it and then `gpu.launch_fun` afterward.
* `gpu.func` has `AffineScope`.
Feel free to comment if you have questions.Thank you.

linuxlonelyeagle committed:
add AffineScope Trait to gpu.launch.

krzysz00 reviewed (commented):
I'll let this sit for people more familiar with the exact semantics here, but this all seems fine to me and if you get no other reviews in a few workdays, ping me and I'll approve

linuxlonelyeagle commented:
> I'll let this sit for people more familiar with the exact semantics here, but this all seems fine to me and if you get no other reviews in a few workdays, ping me and I'll approve

really thank you.

grypp commented:
The PR is also fine for me. I am not familiar with `AffineScope`. Maybe @bondhugula knows?

linuxlonelyeagle committed:
rebase main.

linuxlonelyeagle committed:
rebase main.

linuxlonelyeagle commented:
@bondhugula This PR might need you to review,thank you.

linuxlonelyeagle commented:
@krzysz00  I rebased this PR and reworked the test. Can you approve it? [bondhugula](https://github.com/bondhugula) has given relevant responses to the issues raised in this PR in https://discourse.llvm.org/t/make-threadid-op-is-valid-symbol-introduce-affinesymbol-trait/83702.   Thank you.

linuxlonelyeagle commented:
ping @krzysz00 I think you may have forgotten about this PR.

bondhugula reviewed (approved):
LGTM. Sorry about the delay here.

bondhugula commented on mlir/include/mlir/Dialect/GPU/IR/GPUOps.td:
```diff
@@ -796,7 +796,7 @@ def GPU_LaunchFuncOp :GPU_Op<"launch_func", [
       AutomaticAllocationScope, AttrSizedOperandSegments, GPU_AsyncOpInterface,
       DeclareOpInterfaceMethods<InferIntRangeInterface, ["inferResultRanges"]>,
-      RecursiveMemoryEffects]>,
+      RecursiveMemoryEffects, AffineScope]>,
```
Add in sorted order.

bondhugula commented on mlir/test/Dialect/Affine/ops.mlir:
```diff
@@ -317,13 +317,42 @@ module attributes {gpu.container_module} {
-// CHECK:             affine.for %[[VAL_4:.*]] = %[[VAL_3]] to %[[VAL_2]] step 32 {
-// CHECK:             }
-// CHECK:             gpu.return
+// CHECK-SAME: (%[[VAL_0:.*]]: memref<?x?xf32>) kernel {
+// CHECK: %[[VAL_1:.*]] = arith.constant 1 : index
+// CHECK: %[[VAL_2:.*]] = memref.dim %[[VAL_0]], %[[VAL_1]] : memref<?x?xf32>
+// CHECK: %[[VAL_3:.*]] = arith.constant 0 : index
+// CHECK: affine.for %{{.*}} = %[[VAL_3]] to %[[VAL_2]] step 32 {
+// CHECK: }
+// CHECK: gpu.return
```
Do we need these check lines? I don't think they are testing anything more.

linuxlonelyeagle replied:
Still testing a few things here, tested `memref.dim`, this op does the special processing (forgot the details a bit, sorry)
I seem to understand what you mean though, this `memref.dim` is a `pure` Op, maybe the code could be simplified? (It's the holiday season over here these days, so I might wait a few days to check it out.

linuxlonelyeagle replied:
Bear with me for a moment, I kind of forgot about this test, but it's kind of important and I need a little more time to confirm it.

linuxlonelyeagle replied:
I deleted it.

bondhugula commented on mlir/test/Dialect/Affine/ops.mlir:
```diff
@@ -317,13 +317,42 @@ module attributes {gpu.container_module} {
+
+// -----
+
+// CHECK-LABEL: @gpu_launch_affine
```
Add a summary line on what's being tested for.

linuxlonelyeagle replied:
To be honest with you.Because of the previous changes, the test should not be testing `AffineScope` Trait now.I don't have a great idea for testing the `AffineScope` trait in `gpu.launch` at the moment though.It should now be in test `gpu.launch` to test the changes made in the last PR, and it runs fine (including in my local program)

Introducing `AffineScope` is still necessary though, and I think I can fix this PR first, and I'll go back and add the tests that test `AffineScope` later.

bondhugula commented on mlir/test/Dialect/Affine/ops.mlir:
```diff
@@ -317,13 +317,42 @@ module attributes {gpu.container_module} {
+
+// CHECK-NEXT: %[[VAL_0:.*]] = arith.constant 1 : index
+// CHECK-NEXT: gpu.launch blocks(%[[VAL_1:.*]], %[[VAL_2:.*]], %[[VAL_3:.*]]) in (%[[VAL_4:.*]] = %[[VAL_0]], %[[VAL_5:.*]] = %[[VAL_0]], %[[VAL_6:.*]] = %[[VAL_0]]) threads(%[[VAL_7:.*]], %[[VAL_8:.*]], %[[VAL_9:.*]]) in (%[[VAL_10:.*]] = %[[VAL_0]], %[[VAL_11:.*]] = %[[VAL_0]], %[[VAL_12:.*]] = %[[VAL_0]]) {
+// CHECK-NEXT:   %[[VAL_13:.*]] = gpu.thread_id  x
+// CHECK-NEXT:   %[[VAL_14:.*]] = arith.constant 128 : index
+// CHECK-NEXT:   affine.for %{{.*}} = %[[VAL_13]] to %[[VAL_14]] step 8 {
+// CHECK-NEXT:   }
+// CHECK-NEXT:   gpu.terminator
```
You don't need to match everything. The fact that verification doesn't fail is sufficient. So matching isn't needed.

linuxlonelyeagle committed:
update test and update op define,.

linuxlonelyeagle commented:
I updated the test and `gpu.launch` definitions, is that ok?

linuxlonelyeagle committed:
Merge branch 'main' into add-affinescope-to-gpulaunch
